### PR TITLE
Disable constexpr function body checking in more situations

### PR DESCRIPTION
**Pull request by AaronBallman**
Before C++23, we would check a constexpr function body to diagnose if the function can never be evaluated in a constant expression context. This was previously required standards behavior, but C++23 relaxed the restrictions with P2448R2. While this checking is useful, it is also quite expensive, especially in pathological cases (see #92924 for an example), because it means the mere presence of a constexpr function definition will require constant evaluation even if the function is not used within the TU.

Clang suppresses diagnostics in system headers by default and system headers (like STL implementations) can be full of constexpr function bodies. Now we suppress the check for a diagnostic if the function definition is in a system header or if the `-Winvalid-constexpr` diagnostic is disabled. This should have some mild compile time performance improvements.

Also, the previous implementation would disable the diagnostic in C++23 mode entirely. Due to the benefit of the check, this patch now makes it possible to enable the diagnostic explicitly in C++23 mode.

### CODE REVIEW

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 1]
```diff
@@ -2469,11 +2469,20 @@ static bool CheckConstexprFunctionBody(Sema &SemaRef, const FunctionDecl *Dcl,
   //     base class sub-objects shall be a constexpr constructor.
   //
   // Note that this rule is distinct from the "requirements for a constexpr
-  // function", so is not checked in CheckValid mode.
+  // function", so is not checked in CheckValid mode. Because the check for
+  // constexpr potential is expensive, skip the check if the diagnostic is
+  // disabled, the function is declared in a system header, or we're in C++23
+  // or later mode (see https://wg21.link/P2448).
+  auto SkipCheck = [&SemaRef, Dcl] {
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclCXX.cpp:2476]: Is lambda necessary here? Won't `bool` variable do the job?

Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclCXX.cpp:2476]: Is lambda necessary here?

Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclCXX.cpp:2476]: Is that not to check all this when `Kind != Sema::CheckConstexprKind::Diagnose`? I was under impression that `Expr::isPotentialConstantExpr` is the expensive call.

File: clang/include/clang/Basic/LangOptions.def [hunk 2]
```diff
@@ -505,6 +505,9 @@ COMPATIBLE_LANGOPT(IncrementalExtensions, 1, 0, " True if we want to process sta
 
 BENIGN_LANGOPT(CheckNew, 1, 0, "Do not assume C++ operator new may not return NULL")
 
+BENIGN_LANGOPT(CheckConstexprFunctionBodies, 1, 1,
+               "True if we want to emit a diagnostics for a constexpr function "
```
Comment (erichkeane) [inline clang/include/clang/Basic/LangOptions.def:509]:
```suggestion
               "Emit diagnostics for a constexpr function body that can never be used in a constant expression."
```

Also: I think this is hacky and unfortunate that we're using a bit in `LangOpts` to control a warning's state, which feels kind of dangerous.  I'd love a `FIXME` above this that proposes 1 of 2 ideas: Either find a way for Diagnostics TableGen to be language mode specific, or find a way to put this into CompilerInvocatoin in a way that survives argument rewriting.

File: clang/include/clang/Driver/Options.td [hunk 3]
```diff
@@ -961,6 +961,10 @@ def Wdeprecated : Flag<["-"], "Wdeprecated">, Group<W_Group>,
   HelpText<"Enable warnings for deprecated constructs and define __DEPRECATED">;
 def Wno_deprecated : Flag<["-"], "Wno-deprecated">, Group<W_Group>,
   Visibility<[ClangOption, CC1Option]>;
+def Winvalid_constexpr : Flag<["-"], "Winvalid-constexpr">, Group<W_Group>,
```
Comment (erichkeane) [inline clang/include/clang/Driver/Options.td:964]: I don't think we want these to be separate.  There is a variety of 'Flags' somewhere that auto generates both.

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 1 above)
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclCXX.cpp:2476]: Can this just be a boolean?

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclCXX.cpp]: This is done now.

File: clang/include/clang/Basic/LangOptions.def (hunk 2 above)
Comment (AaronBallman) [reply clang/include/clang/Basic/LangOptions.def]: This is done now.

File: clang/include/clang/Driver/Options.td (hunk 3 above)
Comment (AaronBallman) [reply clang/include/clang/Driver/Options.td]: This was a bit more involved than you might expect, but it's done now and I think it's in a better state as a result.

**Review by erichkeane — APPROVED** [review]

**Pull request merged**